core/threshold/src/networking/tls.rs

```diff
@@ -36,63 +40,173 @@ pub struct ReleasePCRValues {
+        let verifiers = ca_certs
+            .iter()
+            .map(|(subject, ca_cert)| {
+                let mut roots = RootCertStore::empty();
+                roots
+                    .add(CertificateDer::from_slice(&ca_cert.contents))
+                    .map_err(|e| anyhow::anyhow!("{e}"))
+                    .map(|_| Arc::new(roots))
+                    .and_then(|roots| {
+                        WebPkiClientVerifier::builder(roots.clone())
+                            .build()
+                            .map_err(|e| anyhow::anyhow!("{e}"))
+                            .map(|client_verifier| (roots, client_verifier))
+                    })
+                    .and_then(|(roots, client_verifier)| {
+                        WebPkiServerVerifier::builder(roots)
+                            .build()
+                            .map_err(|e| anyhow::anyhow!("{e}"))
+                            .map(|server_verifier| {
+                                (subject.clone(), (client_verifier, server_verifier))
+                            })
+                    })
+            })
+            .collect::<anyhow::Result<HashMap<String, _>>>()?;
+        let root_hint_subjects: Vec<_> = verifiers
+            .values()
+            .flat_map(|(client_verifier, _)| client_verifier.root_hint_subjects())
+            .cloned()
+            .collect();
+        Ok(Self {
+            root_hint_subjects,
+            verifiers,
+            release_pcrs,
+        })
+    }
 }
 
 /// Our custom verifier for our custom mTLS certificates extended with AWS Nitro
 /// attestation documents. It doesn't reimplement normal X.509 certificate
-/// verification and wraps around a well-tested verifier. In addition to the usual X.509 checks, it checks PCR values from the bundled attestation document.
+/// verification and wraps around the well-tested
+/// WebPki[Client|Server]Verifier. In addition to the usual X.509 checks, it
+/// checks PCR values from the bundled attestation document. It also supports
+/// multiple trust root sets configurable at runtime which is handy when working
+/// with multiple MPC contexts.
+///
+/// The TLS certificates are expected to embed the context id in their serial
+/// number. Depending on the context id and the certificate subject name, this
+/// verifier will choose a verifier with just one appropriate CA certificate in
+/// the trust root store to actually verify the certificate.
 #[derive(Debug)]
-pub struct AttestedServerVerifier {
-    verifier: Arc<dyn ServerCertVerifier>,
-    release_pcrs: Arc<Vec<ReleasePCRValues>>,
+pub struct AttestedVerifier {
+    root_hint_subjects: Vec<DistinguishedName>,
+    supported_algs: WebPkiSupportedAlgorithms,
+    // SessionId is supposed to be based on RequestId, and we're representing
+    // ContextId as RequestId so far, so let's say it's all the same for now
+    contexts: RwLock<HashMap<SessionId, Arc<AttestedTLSContext>>>,
+    // are enclave images signed?
```

Comment:
Will there be any security if they are not signed?

Reply:
added a comment in 3e66bc766